### ISSUE TITLE
More consistent deck seeding method?

### ISSUE THREAD

**Issue by lightvector** [Zamiell/hanabi-live#1640]
Currently, the website implements a fixed seed to be a fixed deck order, and all players draw in-order from the shared deck, right? If so, I think it may be possible to improve the seeding for most Hanabi variants, in the sense that the variance between how the seed plays out between different groups should be reduced compared to simply setting the deck order, particularly in games with only 2 or 3 players.

Particularly for 2-player games, and to some extent 3-player games, the way a seed plays out is massively influenced by *who* draws which cards, especially early on. It makes a massive difference if the player with four important but not-yet-playable cards draws the next card which  turns out to be a 5, clogging up their hand, versus a slightly different order of play by the players resulting in someone entirely different drawing the 5. It also can matter significantly who draws the next finessable or bluffable card given the cards, or who draws the duplicate of a card that makes a given hand awkward or not. If all the seed does is set the shuffle order of a common deck that players draw from, then it doesn't preserve this aspect very well.

However, very late in the game, what matters more is exactly how close to the end of the deck the cards that everyone is waiting for are buried, and especially whether the bottom card is the last copy of an earlier discarded card (although mitigated significantly if blind deck bottom plays is on), or whether the second-to-last card is the 2 that was lost earlier, etc. And in that case, the current seeding of just having a common deck works well.

So it seems to me one could reduce seed variance a a little bit more by making it so that:
1. The first, say, ~80% of the deck is pre-dealt between the players such that each player has their *own* draw-pile.
2. When a player draws a card, they draw from their own pile. If their own pile is empty, then they draw from the pile of the player whose remaining draw pile is the largest (deterministically breaking ties by, say, choosing the first player ahead of them in turn order).
3. When all player-specific draw piles are exhausted, the last, say, ~20% of the deck is shared as normal.

This ensures that for a large chunk of the early game, even if players change the order of their play due to different orderings or timings of hints and discards, for the entire early game, the same players will draw the same cards. The same players' hands will get clogged with 5s or not, the same player will always draw the card that is finessable through the player immediately ahead of them, the same players will end up with exactly the same awkward or non-awkward patterns of colors and numbers in their hand if they ultimately make the same plays and discards, even if their timing interleaves differently with other players, etc.

The provision for when a player's draw pile is empty, drawing from the largest other pile - means that even once you get to the mid-to-late game and things have varied a little bit, it will tend to be the case that the same cards show up at roughly similar times in the game, similar to the current seeding. And for the very last portion of the deck, heading into the last few rounds, the ordering of the final cards in the endgame dynamics will be exactly preserved - e.g. the card on the bottom that would kill a given perfect score will always be the same.

The player-specific draw piles and can be purely internal and hidden from any UI, because they don't affect the game from any player's perspective. Within any single game with an unknown seed, such a scheme is completely indistinguishable from just drawing from a shared deck, since if the deck is shuffled uniformly at random then any scheme for how you select the next card to draw will still give you a uniformly random card from those remaining. But it should improve the consistency of the critical early-game situations that come up for shared seeding *between* games and give better comparison between how different playgroups do when faced with that seed.

Thoughts?

**Comment by lightvector** [comment]
(Note: there are also cases where this results in a lot more variance than simply a shared deck, if there is a large discrepancy between two playthroughs for who does all the discarding, or how long it takes someone to discard. But since most draws are from plays, and duplicates aside, drawing the same playable often means the same player will be playing in the future and therefore drawing the next card again, it seems to me that on-average it should be a reduction in variance. I'm not 100% certain though, so I thought it would be interesting to discuss).

**Comment by Zamiell** [comment]
are you a member of the hyphen-ated group? or just a competition player?

**Comment by Zamiell** [comment]
paging mr floris van doorn

**Comment by Zamiell** [comment]
excuse me, dr floris

**Comment by lightvector** [comment]
Heh. Not sure who "dr floris" is, but as for me - I'm a non-hyphen-ated player who has been enjoying the site a lot. Have played a few hundred IRL games with players with independently-developed conventions that aren't anywhere near as extensive as the hyphen-ated ones (and certainly not documented so nicely), but are similar in surprisingly many ways and enough to get perfect scores very often. At point, pondered helping organize a local IRL 2-player Hanabi tournament, which never ended up happening, but the seeding method we came up with was where this idea of splitting a seeded deck into player-specific piles came from.

**Comment by fpvandoorn** [comment]
I have had exactly the same thought to solve the same problem.
The only difference in my idea was that you draw from the bottom of the other player's decks (once your deck runs out), instead of the top. That way the order of the cards is preserved even longer for the other players.

However, I think this method actually increases the variance between different teams. I agree it is important who draws certain cards, but I think it's even more important when a card is drawn. Getting a b2 a whole round earlier (or later) is a big deal, even in the mid or early(ish) game. I think that would make different playthroughs diverge even more quickly.

**Comment by Zamiell** [comment]
floris = fpvandoorn, i brought him up because he mentioned that he had thought deeply about this problem

**Comment by Zamiell** [comment]
also, some people are mentioning that you are the author of KataGo and fireflower. well done! and nice to see you around these parts

**Comment by lightvector** [comment]
@fpvandoorn - Yeah, you might be right, especially for games with many players. Do you think this is still the case for 3 player, or even 2 player? With fewer players it would seem to marginally reduce the effect of a tempo difference on average (due to shorter rounds) and increase the effect of having someone's hand get blocked by the same draw pattern or not, because that player will be taking a much higher proportion of actions and receive a much higher proportion of hints. But maybe the balance is still in favor of trying to preserve tempo?

**Comment by sjdrodge** [comment]
Hope it's okay if I interrupt this thread to just say that analyzing my games carefully with KataGo has helped me improve by ~5 stones since February of this year, and I'd like to take the moment to say thank you, @lightvector. I'm also really delighted to find out that you play Hanabi. Maybe one day we'll convince you to come play with the hyphen-ated group. :)

**Comment by fpvandoorn** [comment]
It's hard to say. With 2-3 players the effect of getting cards later might be smaller, but there are still many hypothetical situations where your suggestion makes game diverge more. Without trying both systems and very carefully analyzing the games, I don't know what effect is larger.

Here is one situations where this suggestion makes games diverge more:
* Suppose that Alice and Bob both have a g4 in their opening hand in a 2p game. Two teams 1 and 2 are playing this deal. In team 1, Alice discards the g4 first, and in team 2 Bob discards the g4 first. Other than that, the game plays out the same. Later in the game for team 1, Bob gets locked for a long time, but the top card of Bob's deck is the card that unlocks Bob. Team 2 never got locked, since they drew the unlocking card on time. With a standard deck - if both teams get the same important cards in the same hands (which is not a given) - this would just mean that Bob gets locked for a couple of turns.

In 2 player games I cannot think of many more examples like that, so maybe you're right that your suggestion improves the similarity of games. However, I don't think that's true for 3+ games, in which the following (or similar) scenarios are quite common:
* Cathy has a g3 near their chop, Alice has the g2 near the top of their deck, and the g1 is in the opening hand. In team 1 Alice might discard first, acceleration their team to the g2, but in team 2 Bob might discard first, and the g2 will be drawn too late, after the g3 is discarded. However, both teams have discarded the same number of times before the g3 was on chop, so with a standard deck they would both get the g2 on time or both lose the g3.

**Comment by Zamiell** [comment]
based on the discussion, i will close this for now

**Comment by Romain672** [comment]
You can totally play around bottom deck risk (bdr)(=discarding a useful card where there is exactly one copy on the deck).
It look pretty hard to do, but here is an idea, the final card of the deck will get a certain value, like 12.
That mean that if any bdr is done when there is <=12 cards in the deck, then the final card will be put at this position.
If it's done when there is >12 cards in the deck, then the final card will stay normal.

And you can do the same for second and third bdr.
That mean there will be seed which require you no bdr to win (1/~40 at the start). But that will not make a game with r3 bdr be win, and another with b3 bdr be lost, which look better for competitive settings if done correctly.
There will be other seed which will make the bdr second last and so will require you to pay lots of attention on the order of some cards.

Still need more ideas about it, but I think if you want a competitive system, that's best.

**Comment by Romain672** [comment]
Here is one possibility to do it: assign a different number to each card of the deck, like (for 10 cards):
r2(2)>r3(8)>r5(7)>r1(10)>r1(1)>b1(9)>r4(5)>r3(3)>b3(4)>r1(6)
Then if any card is played correctly, the deck draw normally r2:
r3(8)>r5(7)>r1(10)>r1(1)>b1(9)>r4(5)>r3(3)>b3(4)>r1(6)
Then, if a bdr of r4 happen, that r4 immediately goes to the lowest number on the deck, those cards are swapped, and the next card is drawn here r3:
r5(7)>r1(10)>**r4(1)**>b1(9)>**r1(5)**>r3(3)>b3(4)>r1(6)

If the lowest number is deep in the deck, that bdr card will be deep too.
